content/en/blog/_posts/advanced-server-side-apply.md

```diff
@@ -0,0 +1,80 @@
+
+Some systems like this continuously reconcile with the cluster, others may only operate when a change is pushed to the git repo. The following considerations are important for both, but more so for the continuously reconciling kind.
+
+CI/CD systems are literally controllers, but for the purpose of APPLY, they are more like users, and unlike other controllers, they need to pay attention to conflicts. Reasoning:
+* Abstractly, CI/CD systems can change anything, which means they could conflict with any controller out there. The recommendation that controllers force conflicts is assuming that controllers change a limited number of things and you can be reasonably sure that they won’t fight with other controllers about those things; that’s clearly not the case for CI/CD controllers.
+* Concrete example: imagine the CI/CD system wants spec.replicas to be 3, because that is checked into source code; but HPA is also enabled for this Deployment, and it thinks there should be 10 replicas. Who should win? We just said that most controllers–including the HPA–should ignore conflicts. The HPA has no idea if it has been enabled incorrectly, and the HPA has no convenient way of informing users of errors.
+* The other common cause of a CI/CD system getting a conflict is probably when it is trying to overwrite a hot-fix (hand-rolled patch) placed there by a system admin / SRE / dev-on-call. You almost certainly don’t want to override that automatically.
+* Of course, sometimes SRE makes an accidental change, or a dev makes an unauthorized change – those you do want to notice and overwrite; however, the CI/CD system can’t tell the difference between these last two cases.
+
+Hopefully this convinces you that CI/CD systems need error paths–a way to back-propagate these conflict errors to humans; in fact, they should have this already, certainly continuous integration systems need some way to report that tests are failing. But maybe we can say something about how humans can deal with errors:
+* Reject the hotfix: the system admin observes the error, and manually force-applies the manifest in question. Then the CI/CD system will be able to apply the manifest successfully and become a co-owner. (Optional: then the system admin applies a blank manifest to relinquish any fields they became a manager for.)
+* Accept the hotfix: the author of the change in question sees the conflict, and edits their change to accept the value running in production.
+* We can also imagine the CI/CD system permitting you to mark a manifest as “force conflicts” somehow– if there’s demand for this we could consider making a more standardized way to do this. A rigorous version of this which lets you declare exactly which conflicts you intend to force would require support from apiserver, but in lieu of that, you can make a second manifest with only that subset of fields.
```

Comment:
I was expecting a different conclusion: the CI/CD system can observe which fields have a different manager (for example, HPA) and can then choose to cede control of that field. I believe SSA has a capacity for this.

The CI/CD system could have hard-coded expectations of which fields (such as `.spec.replicas`) to willingly cede, but other mechanisms are possible (some examples: heuristics; annotations; CRDs).

Up to you if you want to mention this, and a mechanism to surface those warnings / errors is still useful.
Might be worth mentioning that if the CI / CD mechanism is also specified via the Kubernetes API, then reporting a condition via the `.status` of that custom resource object could be useful.

WDYT?

Reply:
Sure, I didn't say anything about that since it's not actually very easy for clients to do right now, but I guess this is supposed to be "advanced" 😅

I will edit

Reply:
I added something, see what you think.